File: python-webrtc-server/webrtc/PerformanceMonitor.py
"""
PerformanceMonitor - Monitors WebRTC performance and resource usage for Python server
Tracks connection metrics, audio processing performance, and system resources
"""

import asyncio
import logging
import psutil
import time
from typing import Dict, List, Optional, Callable, Any
from collections import deque
import json
import threading
from dataclasses import dataclass, asdict

logger = logging.getLogger(__name__)

@dataclass
class PerformanceMetrics:
    """Performance metrics data structure"""
    timestamp: float
    
    # Connection metrics
    active_connections: int = 0
    total_connections: int = 0
    connection_failures: int = 0
    
    # Audio processing metrics
    audio_frames_processed: int = 0
    audio_bytes_processed: int = 0
    audio_processing_latency: float = 0.0
    audio_conversion_errors: int = 0
    
    # Event processing metrics
    events_received: int = 0
    events_sent: int = 0
    event_processing_errors: int = 0
    
    # System resource metrics
    cpu_usage: float = 0.0
    memory_usage: float = 0.0
    memory_usage_mb: float = 0.0
    
    # Network metrics
    bytes_received: int = 0
    bytes_sent: int = 0
    
    # Quality metrics
    packet_loss_rate: float = 0.0
    jitter: float = 0.0
    
    def to_dict(self) -> Dict:
        """Convert to dictionary"""
        return asdict(self)

class PerformanceMonitor:
    """
    Monitors WebRTC server performance and resource usage
    Provides metrics collection, analysis, and optimization recommendations
    """
    
    def __init__(self):
        self.is_monitoring = False
        self.monitoring_task = None
        self.metrics_history: deque = deque(maxlen=1000)  # Keep last 1000 measurements
        
        # Current metrics
        self.current_metrics = PerformanceMetrics(timestamp=time.time())
        
        # Baseline metrics for comparison
        self.baseline_metrics: Optional[PerformanceMetrics] = None
        
        # Performance thresholds
        self.thresholds = {
            'cpu_usage': 80.0,           # %
            'memory_usage': 80.0,        # %
            'audio_latency': 100.0,      # ms
            'packet_loss_rate': 5.0,     # %
            'jitter': 50.0,              # ms
            'connection_failures': 10,    # count per minute
            'processing_errors': 5        # count per minute
        }
        
        # Monitoring configuration
        self.monitoring_interval = 1.0  # seconds
        self.collection_lock = threading.Lock()
        
        # Callbacks
        self.on_metrics_update: Optional[Callable] = None
        self.on_threshold_exceeded: Optional[Callable] = None
        
        # Component references for metrics collection
        self.kvs_master = None
        self.audio_processor = None
        self.event_bridge = None
        
        # Counters for rate calculations
        self.last_metrics_time = time.time()
        self.rate_counters = {
            'connection_failures': 0,
            'processing_errors': 0,
            'audio_frames': 0,
            'events_processed': 0
        }
        
    def set_components(self, kvs_master=None, audio_processor=None, event_bridge=None):
        """
        Set component references for metrics collection
        
        Args:
            kvs_master: KVSWebRTCMaster instance
            audio_processor: AudioProcessor instance
            event_bridge: EventBridge instance
        """
        self.kvs_master = kvs_master
        self.audio_processor = audio_processor
        self.event_bridge = event_bridge
        logger.debug("[PerformanceMonitor] Component references set")
        
    async def start_monitoring(self, interval: float = 1.0):
        """
        Start performance monitoring
        
        Args:
            interval: Monitoring interval in seconds
        """
        if self.is_monitoring:
            logger.warning("[PerformanceMonitor] Already monitoring")
            return
            
        self.monitoring_interval = interval
        self.is_monitoring = True
        
        logger.info(f"[PerformanceMonitor] Starting performance monitoring (interval: {interval}s)")
        
        # Start monitoring task
        self.monitoring_task = asyncio.create_task(self._monitoring_loop())
        
    async def stop_monitoring(self):
        """Stop performance monitoring"""
        if not self.is_monitoring:
            return
            
        logger.debug("[PerformanceMonitor] Stopping performance monitoring")
        
        self.is_monitoring = False
        
        if self.monitoring_task and not self.monitoring_task.done():
            self.monitoring_task.cancel()
            try:
                await self.monitoring_task
            except asyncio.CancelledError:
                pass
                
        logger.debug("[PerformanceMonitor] Performance monitoring stopped")
        
    async def _monitoring_loop(self):
        """Main monitoring loop"""
        try:
            while self.is_monitoring:
                await self.collect_metrics()
                await asyncio.sleep(self.monitoring_interval)
                
        except asyncio.CancelledError:
            logger.debug("[PerformanceMonitor] Monitoring loop cancelled")
        except Exception as e:
            logger.error(f"[PerformanceMonitor] Error in monitoring loop: {e}")
            
    async def collect_metrics(self):
        """Collect all performance metrics"""
        try:
            with self.collection_lock:
                current_time = time.time()
                
                # Create new metrics instance
                metrics = PerformanceMetrics(timestamp=current_time)
                
                # Collect system metrics
                self._collect_system_metrics(metrics)
                
                # Collect WebRTC metrics
                await self._collect_webrtc_metrics(metrics)
                
                # Collect audio processing metrics
                self._collect_audio_metrics(metrics)
                
                # Collect event processing metrics
                self._collect_event_metrics(metrics)
                
                # Calculate rates
                self._calculate_rates(metrics, current_time)
                
                # Update current metrics
                self.current_metrics = metrics
                
                # Store in history
                self.metrics_history.append(metrics)
                
                # Check thresholds
                self._check_thresholds(metrics)
                
                # Notify listeners
                if self.on_metrics_update:
                    await self._safe_callback(self.on_metrics_update, metrics)
                    
        except Exception as e:
            logger.error(f"[PerformanceMonitor] Error collecting metrics: {e}")
            
    def _collect_system_metrics(self, metrics: PerformanceMetrics):
        """Collect system resource metrics"""
        try:
            # CPU usage
            metrics.cpu_usage = psutil.cpu_percent(interval=None)
            
            # Memory usage
            memory = psutil.virtual_memory()
            metrics.memory_usage = memory.percent
            metrics.memory_usage_mb = memory.used / 1024 / 1024
            
            # Network I/O (if available)
            try:
                net_io = psutil.net_io_counters()
                metrics.bytes_received = net_io.bytes_recv
                metrics.bytes_sent = net_io.bytes_sent
            except Exception:
                # Network stats may not be available in all environments
                pass
                
        except Exception as e:
            logger.error(f"[PerformanceMonitor] Error collecting system metrics: {e}")
            
    async def _collect_webrtc_metrics(self, metrics: PerformanceMetrics):
        """Collect WebRTC connection metrics"""
        try:
            if self.kvs_master:
                # Connection metrics
                metrics.active_connections = len(self.kvs_master.get_connected_clients())
                metrics.total_connections = len(self.kvs_master.peer_connections)
                
                # Get connection statistics from KVS master
                # Note: Detailed WebRTC stats would require aiortc peer connection stats
                # This is a simplified implementation
                
        except Exception as e:
            logger.error(f"[PerformanceMonitor] Error collecting WebRTC metrics: {e}")
            
    def _collect_audio_metrics(self, metrics: PerformanceMetrics):
        """Collect audio processing metrics"""
        try:
            if self.audio_processor:
                stats = self.audio_processor.get_processing_stats()
                
                metrics.audio_frames_processed = stats.get('frames_processed', 0)
                metrics.audio_bytes_processed = stats.get('bytes_processed', 0)
                metrics.audio_conversion_errors = stats.get('conversion_errors', 0)
                
                # Calculate processing latency (simplified)
                if stats.get('last_activity'):
                    processing_delay = time.time() - stats['last_activity']
                    metrics.audio_processing_latency = processing_delay * 1000  # Convert to ms
                    
        except Exception as e:
            logger.error(f"[PerformanceMonitor] Error collecting audio metrics: {e}")
            
    def _collect_event_metrics(self, metrics: PerformanceMetrics):
        """Collect event processing metrics"""
        try:
            if self.event_bridge:
                stats = self.event_bridge.get_statistics()
                
                metrics.events_received = stats.get('messages_received', 0)
                metrics.events_sent = stats.get('messages_sent', 0)
                metrics.event_processing_errors = stats.get('processing_errors', 0)
                
        except Exception as e:
            logger.error(f"[PerformanceMonitor] Error collecting event metrics: {e}")
            
    def _calculate_rates(self, metrics: PerformanceMetrics, current_time: float):
        """Calculate rate-based metrics"""
        try:
            time_delta = current_time - self.last_metrics_time
            
            if time_delta > 0:
                # Calculate rates per second
                # Note: This is a simplified rate calculation
                # In production, you might want more sophisticated rate tracking
                pass
                
            self.last_metrics_time = current_time
            
        except Exception as e:
            logger.error(f"[PerformanceMonitor] Error calculating rates: {e}")
            
    def _check_thresholds(self, metrics: PerformanceMetrics):
        """Check if metrics exceed thresholds"""
        violations = []
        
        # Check CPU usage
        if metrics.cpu_usage > self.thresholds['cpu_usage']:
            violations.append({
                'metric': 'cpu_usage',
                'value': metrics.cpu_usage,
                'threshold': self.thresholds['cpu_usage'],
                'severity': 'warning'
            })
            
        # Check memory usage
        if metrics.memory_usage > self.thresholds['memory_usage']:
            violations.append({
                'metric': 'memory_usage',
                'value': metrics.memory_usage,
                'threshold': self.thresholds['memory_usage'],
                'severity': 'warning'
            })
            
        # Check audio processing latency
        if metrics.audio_processing_latency > self.thresholds['audio_latency']:
            violations.append({
                'metric': 'audio_latency',
                'value': metrics.audio_processing_latency,
                'threshold': self.thresholds['audio_latency'],
                'severity': 'warning'
            })
            
        # Check conversion errors
        if metrics.audio_conversion_errors > self.thresholds['processing_errors']:
            violations.append({
                'metric': 'processing_errors',
                'value': metrics.audio_conversion_errors,
                'threshold': self.thresholds['processing_errors'],
                'severity': 'error'
            })
            
        if violations and self.on_threshold_exceeded:
            asyncio.create_task(self._safe_callback(self.on_threshold_exceeded, violations))
            
    async def _safe_callback(self, callback: Callable, *args):
        """Safely invoke callback"""
        try:
            if asyncio.iscoroutinefunction(callback):
                await callback(*args)
            else:
                callback(*args)
        except Exception as e:
            logger.error(f"[PerformanceMonitor] Error in callback: {e}")
            
    def set_baseline(self):
        """Set current metrics as baseline for comparison"""
        self.baseline_metrics = PerformanceMetrics(**asdict(self.current_metrics))
        logger.debug("[PerformanceMonitor] Baseline metrics set")
        
    def compare_with_baseline(self) -> Optional[Dict]:
        """
        Compare current metrics with baseline
        
        Returns:
            Dictionary with comparison results or None if no baseline
        """
        if not self.baseline_metrics:
            return None
            
        comparison = {}
        current_dict = asdict(self.current_metrics)
        baseline_dict = asdict(self.baseline_metrics)
        
        for key, current_value in current_dict.items():
            if isinstance(current_value, (int, float)) and key in baseline_dict:
                baseline_value = baseline_dict[key]
                if baseline_value != 0:
                    difference = current_value - baseline_value
                    percent_change = (difference / baseline_value) * 100
                    
                    comparison[key] = {
                        'current': current_value,
                        'baseline': baseline_value,
                        'difference': difference,
                        'percent_change': round(percent_change, 2)
                    }
                    
        return comparison
        
    def get_metrics_summary(self) -> Dict:
        """
        Get comprehensive metrics summary
        
        Returns:
            Dictionary with metrics summary
        """
        # Calculate averages over last 30 seconds
        recent_time = time.time() - 30
        recent_metrics = [m for m in self.metrics_history if m.timestamp >= recent_time]
        
        averages = {}
        if recent_metrics:
            metrics_dicts = [asdict(m) for m in recent_metrics]
            for key in asdict(self.current_metrics).keys():
                if key != 'timestamp':
                    values = [m[key] for m in metrics_dicts if isinstance(m[key], (int, float))]
                    if values:
                        averages[key] = sum(values) / len(values)
                        
        return {
            'current': asdict(self.current_metrics),
            'baseline': asdict(self.baseline_metrics) if self.baseline_metrics else None,
            'comparison': self.compare_with_baseline(),
            'averages': averages,
            'history_size': len(self.metrics_history),
            'is_monitoring': self.is_monitoring,
            'thresholds': self.thresholds
        }
        
    def get_performance_report(self) -> Dict:
        """
        Get detailed performance report with recommendations
        
        Returns:
            Comprehensive performance report
        """
        summary = self.get_metrics_summary()
        
        return {
            **summary,
            'health_score': self.calculate_health_score(),
            'recommendations': self.get_recommendations(),
            'optimization_suggestions': self.get_optimization_suggestions()
        }
        
    def calculate_health_score(self) -> float:
        """
        Calculate overall system health score (0-100)
        
        Returns:
            Health score as percentage
        """
        score = 100.0
        
        # Deduct points for threshold violations
        if self.current_metrics.cpu_usage > self.thresholds['cpu_usage']:
            score -= 20
            
        if self.current_metrics.memory_usage > self.thresholds['memory_usage']:
            score -= 15
            
        if self.current_metrics.audio_processing_latency > self.thresholds['audio_latency']:
            score -= 25
            
        if self.current_metrics.audio_conversion_errors > self.thresholds['processing_errors']:
            score -= 30
            
        if self.current_metrics.event_processing_errors > self.thresholds['processing_errors']:
            score -= 20
            
        # Connection health
        if self.current_metrics.active_connections == 0 and self.current_metrics.total_connections > 0:
            score -= 40  # Connections exist but none are active
            
        return max(0.0, min(100.0, score))
        
    def get_recommendations(self) -> List[Dict]:
        """
        Get performance recommendations based on current metrics
        
        Returns:
            List of recommendation dictionaries
        """
        recommendations = []
        
        if self.current_metrics.cpu_usage > self.thresholds['cpu_usage']:
            recommendations.append({
                'type': 'cpu',
                'severity': 'warning',
                'message': f'High CPU usage: {self.current_metrics.cpu_usage:.1f}%',
                'action': 'Consider optimizing audio processing or reducing concurrent connections'
            })
            
        if self.current_metrics.memory_usage > self.thresholds['memory_usage']:
            recommendations.append({
                'type': 'memory',
                'severity': 'warning',
                'message': f'High memory usage: {self.current_metrics.memory_usage:.1f}%',
                'action': 'Check for memory leaks and optimize buffer management'
            })
            
        if self.current_metrics.audio_processing_latency > self.thresholds['audio_latency']:
            recommendations.append({
                'type': 'latency',
                'severity': 'error',
                'message': f'High audio processing latency: {self.current_metrics.audio_processing_latency:.1f}ms',
                'action': 'Optimize audio processing pipeline and check system load'
            })
            
        if self.current_metrics.audio_conversion_errors > 0:
            recommendations.append({
                'type': 'audio_errors',
                'severity': 'error',
                'message': f'Audio conversion errors detected: {self.current_metrics.audio_conversion_errors}',
                'action': 'Check audio format compatibility and error handling'
            })
            
        return recommendations
        
    def get_optimization_suggestions(self) -> List[Dict]:
        """
        Get optimization suggestions based on performance analysis
        
        Returns:
            List of optimization suggestion dictionaries
        """
        suggestions = []
        
        # Analyze connection patterns
        if self.current_metrics.active_connections > 10:
            suggestions.append({
                'category': 'scaling',
                'suggestion': 'Consider implementing connection pooling for high concurrent usage',
                'impact': 'medium',
                'effort': 'high'
            })
            
        # Analyze audio processing
        if self.current_metrics.audio_processing_latency > 50:
            suggestions.append({
                'category': 'audio',
                'suggestion': 'Implement audio processing optimization or use hardware acceleration',
                'impact': 'high',
                'effort': 'medium'
            })
            
        # Analyze error rates
        if self.current_metrics.audio_conversion_errors > 0:
            suggestions.append({
                'category': 'reliability',
                'suggestion': 'Implement better error handling and recovery mechanisms',
                'impact': 'high',
                'effort': 'low'
            })
            
        # Memory optimization
        if self.current_metrics.memory_usage_mb > 500:
            suggestions.append({
                'category': 'memory',
                'suggestion': 'Implement memory optimization and garbage collection tuning',
                'impact': 'medium',
                'effort': 'medium'
            })
            
        return suggestions
        
    def export_metrics(self, format: str = 'json') -> str:
        """
        Export metrics data for analysis
        
        Args:
            format: Export format ('json' or 'csv')
            
        Returns:
            Exported data as string
        """
        if format == 'json':
            export_data = {
                'summary': self.get_metrics_summary(),
                'full_history': [asdict(m) for m in self.metrics_history],
                'export_time': time.time()
            }
            return json.dumps(export_data, indent=2)
        else:
            raise ValueError(f"Unsupported export format: {format}")
            
    def set_thresholds(self, new_thresholds: Dict):
        """
        Update performance thresholds
        
        Args:
            new_thresholds: Dictionary with new threshold values
        """
        self.thresholds.update(new_thresholds)
        logger.debug(f"[PerformanceMonitor] Updated thresholds: {self.thresholds}")
        
    def record_connection_failure(self):
        """Record a connection failure for metrics"""
        self.rate_counters['connection_failures'] += 1
        
    def record_processing_error(self):
        """Record a processing error for metrics"""
        self.rate_counters['processing_errors'] += 1
        
    async def cleanup(self):
        """Clean up monitoring resources"""
        await self.stop_monitoring()
        self.metrics_history.clear()
        logger.debug("[PerformanceMonitor] Cleanup completed")

File: react-webrtc-client/src/helper/PerformanceMonitor.js
/**
 * PerformanceMonitor - Monitors WebRTC performance and resource usage
 * Tracks audio latency, connection quality, and resource consumption
 */

class PerformanceMonitor {
    constructor() {
        this.isMonitoring = false;
        this.monitoringInterval = null;
        this.metricsHistory = [];
        this.maxHistorySize = 1000; // Keep last 1000 measurements
        
        // Performance metrics
        this.metrics = {
            // Audio metrics
            audioLatency: 0,
            audioPacketsLost: 0,
            audioJitter: 0,
            audioLevel: 0,
            
            // Connection metrics
            connectionState: 'disconnected',
            iceConnectionState: 'new',
            bytesReceived: 0,
            bytesSent: 0,
            packetsReceived: 0,
            packetsSent: 0,
            
            // Resource metrics
            cpuUsage: 0,
            memoryUsage: 0,
            
            // Timing metrics
            connectionTime: 0,
            firstAudioTime: 0,
            
            // Error metrics
            connectionErrors: 0,
            audioErrors: 0,
            dataChannelErrors: 0
        };
        
        // Baseline measurements for comparison
        this.baseline = null;
        
        // Performance thresholds
        this.thresholds = {
            audioLatency: 200, // ms
            audioJitter: 50,   // ms
            packetLoss: 5,     // %
            cpuUsage: 80,      // %
            memoryUsage: 500   // MB
        };
        
        // Callbacks
        this.onMetricsUpdate = null;
        this.onThresholdExceeded = null;
        
        // Connection timing
        this.connectionStartTime = null;
        this.firstAudioReceived = false;
        this.firstAudioTime = null;
    }

    /**
     * Start performance monitoring
     * @param {RTCPeerConnection} peerConnection - WebRTC peer connection
     * @param {Object} options - Monitoring options
     */
    startMonitoring(peerConnection, options = {}) {
        if (this.isMonitoring) {
            console.warn('[PerformanceMonitor] Already monitoring');
            return;
        }

        this.peerConnection = peerConnection;
        this.isMonitoring = true;
        this.connectionStartTime = Date.now();
        
        const interval = options.interval || 1000; // Default 1 second
        
        console.log(`[PerformanceMonitor] Starting performance monitoring with interval ${interval}ms`);
        
        this.monitoringInterval = setInterval(() => {
            this.collectMetrics();
        }, interval);
        
        console.log(`[PerformanceMonitor] Monitoring interval created: ${this.monitoringInterval}`);
        
        // Set up connection state monitoring
        this.setupConnectionMonitoring();
    }

    /**
     * Stop performance monitoring
     */
    stopMonitoring() {
        if (!this.isMonitoring) {
            console.log('[PerformanceMonitor] Already stopped, ignoring stopMonitoring call');
            return;
        }

        console.log('[PerformanceMonitor] Stopping performance monitoring');
        
        this.isMonitoring = false;
        
        if (this.monitoringInterval) {
            console.log('[PerformanceMonitor] Clearing monitoring interval');
            clearInterval(this.monitoringInterval);
            this.monitoringInterval = null;
        } else {
            console.log('[PerformanceMonitor] No monitoring interval to clear');
        }
        
        this.peerConnection = null;
        console.log('[PerformanceMonitor] Performance monitoring stopped successfully');
    }

    /**
     * Set up connection state monitoring
     */
    setupConnectionMonitoring() {
        if (!this.peerConnection) return;

        // Monitor connection state changes
        const originalOnConnectionStateChange = this.peerConnection.onconnectionstatechange;
        this.peerConnection.onconnectionstatechange = () => {
            this.metrics.connectionState = this.peerConnection.connectionState;
            
            if (this.peerConnection.connectionState === 'connected' && this.connectionStartTime) {
                this.metrics.connectionTime = Date.now() - this.connectionStartTime;
            }
            
            if (originalOnConnectionStateChange) {
                originalOnConnectionStateChange();
            }
        };

        // Monitor ICE connection state
        const originalOnIceConnectionStateChange = this.peerConnection.oniceconnectionstatechange;
        this.peerConnection.oniceconnectionstatechange = () => {
            this.metrics.iceConnectionState = this.peerConnection.iceConnectionState;
            
            if (originalOnIceConnectionStateChange) {
                originalOnIceConnectionStateChange();
            }
        };
    }

    /**
     * Collect performance metrics
     */
    async collectMetrics() {
        if (!this.peerConnection || !this.isMonitoring) {
            return;
        }

        try {
            // Get WebRTC statistics
            const stats = await this.peerConnection.getStats();
            this.processWebRTCStats(stats);
            
            // Get system resource usage
            await this.collectResourceMetrics();
            
            // Calculate derived metrics
            this.calculateDerivedMetrics();
            
            // Store metrics in history
            this.storeMetrics();
            
            // Check thresholds
            this.checkThresholds();
            
            // Notify listeners
            if (this.onMetricsUpdate) {
                this.onMetricsUpdate(this.getMetricsSummary());
            }
            
        } catch (error) {
            console.error('[PerformanceMonitor] Error collecting metrics:', error);
        }
    }

    /**
     * Process WebRTC statistics
     * @param {RTCStatsReport} stats - WebRTC stats report
     */
    processWebRTCStats(stats) {
        let audioInboundStats = null;
        let audioOutboundStats = null;
        let candidatePairStats = null;

        // Find relevant stats
        stats.forEach(stat => {
            if (stat.type === 'inbound-rtp' && stat.kind === 'audio') {
                audioInboundStats = stat;
            } else if (stat.type === 'outbound-rtp' && stat.kind === 'audio') {
                audioOutboundStats = stat;
            } else if (stat.type === 'candidate-pair' && stat.state === 'succeeded') {
                candidatePairStats = stat;
            }
        });

        // Process audio inbound stats
        if (audioInboundStats) {
            this.metrics.packetsReceived = audioInboundStats.packetsReceived || 0;
            this.metrics.bytesReceived = audioInboundStats.bytesReceived || 0;
            this.metrics.audioPacketsLost = audioInboundStats.packetsLost || 0;
            this.metrics.audioJitter = (audioInboundStats.jitter || 0) * 1000; // Convert to ms
            
            // Track first audio received
            if (!this.firstAudioReceived && this.metrics.packetsReceived > 0) {
                this.firstAudioReceived = true;
                this.firstAudioTime = Date.now();
                if (this.connectionStartTime) {
                    this.metrics.firstAudioTime = this.firstAudioTime - this.connectionStartTime;
                }
            }
        }

        // Process audio outbound stats
        if (audioOutboundStats) {
            this.metrics.packetsSent = audioOutboundStats.packetsSent || 0;
            this.metrics.bytesSent = audioOutboundStats.bytesSent || 0;
        }

        // Process candidate pair stats for RTT
        if (candidatePairStats) {
            this.metrics.audioLatency = candidatePairStats.currentRoundTripTime * 1000 || 0; // Convert to ms
        }
    }

    /**
     * Collect system resource metrics
     */
    async collectResourceMetrics() {
        try {
            // Memory usage (if available)
            if (performance.memory) {
                this.metrics.memoryUsage = Math.round(performance.memory.usedJSHeapSize / 1024 / 1024); // MB
            }

            // CPU usage estimation (simplified)
            // Note: Actual CPU usage is not directly available in browsers
            // This is a rough estimation based on timing
            const startTime = performance.now();
            await new Promise(resolve => setTimeout(resolve, 0));
            const endTime = performance.now();
            
            // Very rough CPU usage estimation
            const timingDelta = endTime - startTime;
            this.metrics.cpuUsage = Math.min(100, Math.max(0, timingDelta * 10));
            
        } catch (error) {
            console.error('[PerformanceMonitor] Error collecting resource metrics:', error);
        }
    }

    /**
     * Calculate derived metrics
     */
    calculateDerivedMetrics() {
        // Calculate packet loss percentage
        const totalPackets = this.metrics.packetsReceived + this.metrics.audioPacketsLost;
        if (totalPackets > 0) {
            this.metrics.packetLossPercentage = (this.metrics.audioPacketsLost / totalPackets) * 100;
        } else {
            this.metrics.packetLossPercentage = 0;
        }
        
        // Add timestamp
        this.metrics.timestamp = Date.now();
    }

    /**
     * Store metrics in history
     */
    storeMetrics() {
        const metricsSnapshot = { ...this.metrics };
        this.metricsHistory.push(metricsSnapshot);
        
        // Limit history size
        if (this.metricsHistory.length > this.maxHistorySize) {
            this.metricsHistory.shift();
        }
    }

    /**
     * Check performance thresholds
     */
    checkThresholds() {
        const violations = [];
        
        if (this.metrics.audioLatency > this.thresholds.audioLatency) {
            violations.push({
                metric: 'audioLatency',
                value: this.metrics.audioLatency,
                threshold: this.thresholds.audioLatency,
                severity: 'warning'
            });
        }
        
        if (this.metrics.audioJitter > this.thresholds.audioJitter) {
            violations.push({
                metric: 'audioJitter',
                value: this.metrics.audioJitter,
                threshold: this.thresholds.audioJitter,
                severity: 'warning'
            });
        }
        
        if (this.metrics.packetLossPercentage > this.thresholds.packetLoss) {
            violations.push({
                metric: 'packetLoss',
                value: this.metrics.packetLossPercentage,
                threshold: this.thresholds.packetLoss,
                severity: 'error'
            });
        }
        
        if (this.metrics.cpuUsage > this.thresholds.cpuUsage) {
            violations.push({
                metric: 'cpuUsage',
                value: this.metrics.cpuUsage,
                threshold: this.thresholds.cpuUsage,
                severity: 'warning'
            });
        }
        
        if (this.metrics.memoryUsage > this.thresholds.memoryUsage) {
            violations.push({
                metric: 'memoryUsage',
                value: this.metrics.memoryUsage,
                threshold: this.thresholds.memoryUsage,
                severity: 'warning'
            });
        }
        
        if (violations.length > 0 && this.onThresholdExceeded) {
            this.onThresholdExceeded(violations);
        }
    }

    /**
     * Set baseline measurements for comparison
     */
    setBaseline() {
        this.baseline = { ...this.metrics };
        console.log('[PerformanceMonitor] Baseline set:', this.baseline);
    }

    /**
     * Compare current metrics with baseline
     * @returns {Object} Comparison results
     */
    compareWithBaseline() {
        if (!this.baseline) {
            return null;
        }

        const comparison = {};
        
        Object.keys(this.metrics).forEach(key => {
            if (typeof this.metrics[key] === 'number' && typeof this.baseline[key] === 'number') {
                const current = this.metrics[key];
                const baseline = this.baseline[key];
                const difference = current - baseline;
                const percentChange = baseline !== 0 ? (difference / baseline) * 100 : 0;
                
                comparison[key] = {
                    current,
                    baseline,
                    difference,
                    percentChange: Math.round(percentChange * 100) / 100
                };
            }
        });
        
        return comparison;
    }

    /**
     * Get metrics summary
     * @returns {Object} Metrics summary
     */
    getMetricsSummary() {
        return {
            current: { ...this.metrics },
            baseline: this.baseline ? { ...this.baseline } : null,
            comparison: this.compareWithBaseline(),
            history: this.metricsHistory.slice(-10), // Last 10 measurements
            isMonitoring: this.isMonitoring
        };
    }

    /**
     * Get performance report
     * @returns {Object} Detailed performance report
     */
    getPerformanceReport() {
        const summary = this.getMetricsSummary();
        
        // Calculate averages over last 30 seconds
        const recentMetrics = this.metricsHistory.filter(
            m => Date.now() - m.timestamp < 30000
        );
        
        const averages = {};
        if (recentMetrics.length > 0) {
            Object.keys(this.metrics).forEach(key => {
                if (typeof this.metrics[key] === 'number') {
                    const values = recentMetrics.map(m => m[key]).filter(v => typeof v === 'number');
                    if (values.length > 0) {
                        averages[key] = values.reduce((sum, val) => sum + val, 0) / values.length;
                    }
                }
            });
        }
        
        return {
            ...summary,
            averages,
            recommendations: this.getRecommendations(),
            healthScore: this.calculateHealthScore()
        };
    }

    /**
     * Calculate overall health score (0-100)
     * @returns {number} Health score
     */
    calculateHealthScore() {
        let score = 100;
        
        // Deduct points for threshold violations
        if (this.metrics.audioLatency > this.thresholds.audioLatency) {
            score -= 20;
        }
        
        if (this.metrics.packetLossPercentage > this.thresholds.packetLoss) {
            score -= 30;
        }
        
        if (this.metrics.audioJitter > this.thresholds.audioJitter) {
            score -= 15;
        }
        
        if (this.metrics.cpuUsage > this.thresholds.cpuUsage) {
            score -= 10;
        }
        
        if (this.metrics.memoryUsage > this.thresholds.memoryUsage) {
            score -= 10;
        }
        
        // Connection state penalties
        if (this.metrics.connectionState !== 'connected') {
            score -= 50;
        }
        
        return Math.max(0, Math.min(100, score));
    }

    /**
     * Get performance recommendations
     * @returns {Array} Array of recommendation objects
     */
    getRecommendations() {
        const recommendations = [];
        
        if (this.metrics.audioLatency > this.thresholds.audioLatency) {
            recommendations.push({
                type: 'latency',
                severity: 'warning',
                message: 'High audio latency detected. Consider checking network conditions.',
                action: 'Check network connection and consider using a wired connection.'
            });
        }
        
        if (this.metrics.packetLossPercentage > this.thresholds.packetLoss) {
            recommendations.push({
                type: 'packet_loss',
                severity: 'error',
                message: 'Significant packet loss detected. Audio quality may be affected.',
                action: 'Check network stability and bandwidth availability.'
            });
        }
        
        if (this.metrics.audioJitter > this.thresholds.audioJitter) {
            recommendations.push({
                type: 'jitter',
                severity: 'warning',
                message: 'High audio jitter detected. May cause audio quality issues.',
                action: 'Ensure stable network connection and close bandwidth-intensive applications.'
            });
        }
        
        if (this.metrics.cpuUsage > this.thresholds.cpuUsage) {
            recommendations.push({
                type: 'cpu',
                severity: 'warning',
                message: 'High CPU usage detected. May affect performance.',
                action: 'Close unnecessary applications and browser tabs.'
            });
        }
        
        if (this.metrics.memoryUsage > this.thresholds.memoryUsage) {
            recommendations.push({
                type: 'memory',
                severity: 'warning',
                message: 'High memory usage detected.',
                action: 'Close unnecessary browser tabs and applications.'
            });
        }
        
        return recommendations;
    }

    /**
     * Export metrics data for analysis
     * @returns {Object} Exportable metrics data
     */
    exportMetrics() {
        return {
            summary: this.getMetricsSummary(),
            fullHistory: this.metricsHistory,
            thresholds: this.thresholds,
            exportTime: new Date().toISOString()
        };
    }

    /**
     * Set custom thresholds
     * @param {Object} newThresholds - New threshold values
     */
    setThresholds(newThresholds) {
        this.thresholds = { ...this.thresholds, ...newThresholds };
        console.log('[PerformanceMonitor] Updated thresholds:', this.thresholds);
    }

    /**
     * Record audio level for monitoring
     * @param {number} level - Audio level (0-100)
     */
    recordAudioLevel(level) {
        this.metrics.audioLevel = level;
    }

    /**
     * Record error for tracking
     * @param {string} errorType - Type of error
     */
    recordError(errorType) {
        switch (errorType) {
            case 'connection':
                this.metrics.connectionErrors++;
                break;
            case 'audio':
                this.metrics.audioErrors++;
                break;
            case 'datachannel':
                this.metrics.dataChannelErrors++;
                break;
            default:
                console.warn(`Unknown error type: ${errorType}`);
                break;
        }
    }
}

export default PerformanceMonitor;